src/main/java/bio/terra/workspace/service/workspace/flight/CreateCustomGcpRolesStep.java

```diff
@@ -0,0 +1,56 @@
+package bio.terra.workspace.service.workspace.flight;
+
+import static bio.terra.workspace.service.workspace.flight.WorkspaceFlightMapKeys.GOOGLE_PROJECT_ID;
+
+import bio.terra.cloudres.google.iam.IamCow;
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.Step;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.exception.RetryException;
+import bio.terra.workspace.service.iam.CustomGcpIamRole;
+import bio.terra.workspace.service.iam.CustomGcpIamRoleMapping;
+import com.google.api.services.iam.v1.model.CreateRoleRequest;
+import com.google.api.services.iam.v1.model.Role;
+import java.io.IOException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CreateCustomGcpRolesStep implements Step {
+
+  private final IamCow iamCow;
+
+  private final Logger logger = LoggerFactory.getLogger(CreateCustomGcpRolesStep.class);
+
+  public CreateCustomGcpRolesStep(IamCow iamCow) {
+    this.iamCow = iamCow;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext flightContext)
+      throws InterruptedException, RetryException {
+    String projectId = flightContext.getWorkingMap().get(GOOGLE_PROJECT_ID, String.class);
+    for (CustomGcpIamRole customRole : CustomGcpIamRoleMapping.customIamRoles) {
+      try {
+        Role gcpRole = new Role().setIncludedPermissions(customRole.getIncludedPermissions());
+        CreateRoleRequest request =
+            new CreateRoleRequest().setRole(gcpRole).setRoleId(customRole.getRoleName());
+        logger.info(
+            "Creating role {} with permissions {} in project {}",
+            customRole.getRoleName(),
+            customRole.getIncludedPermissions(),
+            projectId);
+        iamCow.projects().roles().create("projects/" + projectId, request).execute();
+      } catch (IOException e) {
+        // Retry on IO exceptions thrown by CRL.
+        throw new RetryException(e);
+      }
+    }
+    return StepResult.getStepResultSuccess();
+  }
+
+  @Override
+  public StepResult undoStep(FlightContext flightContext) throws InterruptedException {
+    // No need to delete roles if the project is being deleted.
```

Comment:
How do you know the project is being deleted? If this step is reused somewhere else, will that logic hold?

Reply:
We know the project is being deleted because of other steps in this flight. If this step is reused that might not hold, but I don't know where else we would need this step. Updated the comment, but I don't think it's worth cleaning up the custom roles if we don't need to.